packages/workflow/test/Workflow.test.ts

```diff
@@ -2044,4 +2020,368 @@ describe('Workflow', () => {
+			expect(result).toBe(set1Node);
+		});
+
+		test('should traverse through non-main connections to find main input', () => {
+			const set1Node = WORKFLOW_WITH_MIXED_CONNECTIONS.getNode('Set1')!;
+			const result = WORKFLOW_WITH_MIXED_CONNECTIONS.getParentMainInputNode(set1Node);
+			expect(result).toBe(set1Node);
+		});
+
+		test('should handle nested non-main connections', () => {
+			const set1Node = WORKFLOW_WITH_LOOPS.getNode('Set1')!;
+			const result = WORKFLOW_WITH_LOOPS.getParentMainInputNode(set1Node);
+			expect(result).toBe(set1Node);
+		});
+	});
+
+	describe('getNodeConnectionIndexes', () => {
+		test('should return undefined for non-existent parent node', () => {
+			const result = SIMPLE_WORKFLOW.getNodeConnectionIndexes('Set', 'NonExistentNode');
+			expect(result).toBeUndefined();
+		});
+
+		test('should return undefined for nodes without connections', () => {
+			const result = SIMPLE_WORKFLOW.getNodeConnectionIndexes('Start', 'Set1');
+			expect(result).toBeUndefined();
+		});
+
+		test('should return correct connection indexes for direct connections', () => {
+			const result = SIMPLE_WORKFLOW.getNodeConnectionIndexes('Set', 'Start');
+			expect(result).toEqual({
+				sourceIndex: 0,
+				destinationIndex: 0,
+			});
+		});
+
+		test('should return correct connection indexes for multi-step connections', () => {
+			const result = SIMPLE_WORKFLOW.getNodeConnectionIndexes('Set1', 'Start');
+			expect(result).toEqual({
+				sourceIndex: 0,
+				destinationIndex: 0,
+			});
+		});
+
+		test('should return undefined when depth is 0', () => {
+			const result = SIMPLE_WORKFLOW.getNodeConnectionIndexes(
+				'Set',
+				'Start',
+				NodeConnectionType.Main,
+				0,
+			);
+			expect(result).toBeUndefined();
+		});
+
+		test('should handle workflows with multiple connection indexes', () => {
+			const result = WORKFLOW_WITH_SWITCH.getNodeConnectionIndexes('Set', 'Switch');
+			expect(result).toEqual({
+				sourceIndex: 1,
+				destinationIndex: 0,
+			});
+		});
+	});
+
+	describe('getStartNode', () => {
+		const manualTriggerNode = mock<INode>({
+			name: 'ManualTrigger',
+			type: 'n8n-nodes-base.manualTrigger',
+		});
+		const scheduleTriggerNode = mock<INode>({
+			name: 'ScheduleTrigger',
+			type: 'n8n-nodes-base.scheduleTrigger',
+		});
+		const httpRequestNode = mock<INode>({
+			name: 'HTTP Request',
+			type: 'n8n-nodes-base.httpRequest',
+		});
+		const set1Node = mock<INode>({
+			name: 'Set1',
+			type: 'n8n-nodes-base.set',
+		});
+		const disabledSetNode = mock<INode>({
+			name: 'Set Disabled',
+			type: 'n8n-nodes-base.set',
+			disabled: true,
+		});
+
+		test('returns first trigger node when multiple start nodes exist', () => {
+			const workflow = new Workflow({
+				nodes: [manualTriggerNode, scheduleTriggerNode],
+				connections: {},
+				active: false,
+				nodeTypes,
+			});
+
+			expect(workflow.getStartNode()).toBe(scheduleTriggerNode);
```

Comment:
Why is scheduleTriggerNode the "first"? manualTriggerNode is earlier in the array

Reply:
I did not verify the correctness of these because for now I just want to have tests cover the existing implementation, so that we can get to a high enough coverage, after which we can start refactoring this code.